### PR TITLE
feat: salt prefix support

### PR DESCRIPTION
**Pull request by bemasc**
This is an **experimental** version of Outline with "custom salt prefix" support, currently working on Windows, Linux, and Android.  It has not yet been tested on macOS/iOS.  A [pre-built APK](https://github.com/Jigsaw-Code/outline-client/raw/c4775ecc92999edc5841f02ce4ba35f7c8f61f1a/app-debug.apk) is available for Android.

Custom salt prefixes are **experimental** and are **not as secure** as standard Outline, so this version should be used **only for testing**.

To use the prefix feature with [Dynamic Keys](https://www.reddit.com/r/outlinevpn/wiki/index/dynamic_access_keys/), add a `"prefix"` key to the JSON object, with a string value representing the prefix you want.  (You can use escape codes like `\xFF` to represent non-printable byte values.)  The prefix is a freeform string of up to 32 bytes, each represented as a unicode codepoint in the U+0 to U+FF range.

To use this feature with `ss://` Access Keys, add a URL parameter like `&prefix=...` to the ss:// URL's query parameters, so that the URL looks like `.../?outline=1&prefix=OUTLINE#Name-of-the-server`.  You must use the `encodeURIComponent()` function to convert your prefix into escaped form suitable for inclusion in a URL.

Possible prefixes of interest (note: longer prefixes are less secure):

| JSON | URI-encoded|
|------------------|---------|
| `"POST "` | `POST%20` |
| `"HTTP/1.1"` | `HTTP%2F1.1` |
| `"POST / HTTP/1.1"` | `POST%20%2F%20HTTP%2F1.1` |
| `"GET / HTTP/1.1\r\n\r\n"` | `GET%20%2F%20HTTP%2F1.1%0D%0A%0D%0A` |
| [`"\x16\x03\x01\x02\x00"`](https://geneva.cs.umd.edu/posts/iran-whitelister/#:~:text=Example%3A%20GET%20testing123-,HTTPS%20Fingerprint,-The%20HTTPS%20protocol) | `%16%03%01%02%00` |

More potentially useful prefixes are mentioned [here](https://geneva.cs.umd.edu/posts/iran-whitelister/).  Note that some prefixes may be more effective on specific port numbers.

### CODE REVIEW

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:225]: Perhaps we should log the exception, since we are losing that information.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:176]: Can we get rid of `TunnelConfig` and `makeTunnelConfig`?

Ideally we would stick to the JSONObject (perhaps proto in the future) and treat it as opaque in the platform-specific code.

File: cordova-plugin-outline/android/aidl/org/outline/shadowsocks/ShadowsocksConfig.aidl
Comment (fortuna) [inline cordova-plugin-outline/android/aidl/org/outline/shadowsocks/ShadowsocksConfig.aidl:22]: @jyyi1 We'll probably be able to get rid of this if we share protocol buffers across implementations

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:222]: It would be helpful to provide a Client factory that takes a JSON object instead (perhaps proto in the future), so the Android implementation doesn't have to know about the fields.

File: src/www/app/outline_server_repository/access_key_serialization.ts
Comment (fortuna) [inline src/www/app/outline_server_repository/access_key_serialization.ts:42]: Let's have an explicit prefix field instead. The dynamic config should align with the SessionConfig.

File: src/electron/go_vpn_tunnel.ts
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:293]: It would be nice to have a single `-proxyConfig` flag with the json object, so we don't need to know about the fields here.

**Review by fortuna — COMMENTED** [review]
I'm impressed by how quickly you got this together. Thanks!

I have some comments on how we can decouple the platform-specific code from the protocol details.

**Review by fortuna — COMMENTED** [review]
Why are the Apple builds hitting an Android build error?

File: src/www/app/outline_server_repository/index.ts
Comment (fortuna) [inline src/www/app/outline_server_repository/index.ts:40]: People are complaining that the app is saying keys with the prefix `prefix=%16%03%01%40%00%01` are invalid. I think there's something wrong with the serialization. Maybe that's not valid utf8/unicode? We need some unit testing.

Comment (fortuna) [reply src/www/app/outline_server_repository/index.ts:40]: Note that it works with the prod app (I guess we just ignore the prefix), but not the apk you provided. I can give you a key in private.

**Review by fortuna — CHANGES_REQUESTED** [review]

**Comment by bemasc** [comment]
This code now builds for Mac and iOS.  (I haven't been able to test yet to confirm that it works.)

> Why are the Apple builds hitting an Android build error?

It looks like this was because the `platforms/android/` directory existed, because I had checked in the APK.  I moved the APK to the root.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:226]: Done

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:176]: I think that's a good idea, but I don't want to combine that with this change.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:222]: Agreed.  However, given the large change required in each platform's code, I don't think that should go in this PR.

File: src/www/app/outline_server_repository/index.ts
Comment (bemasc) [reply src/www/app/outline_server_repository/index.ts:40]: I've updated the instructions to help people avoid writing invalid prefixes.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java:131]: Note for future: we can probably get rid of the isUdpEnabled option by passing a Client that knows it needs to manipulate the DNS requests.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:177]: Don't you need to check if prefix is present?

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:260]: Note for future: at this point perhaps the UDP handler could be wrapped to handle the UDP unavailability.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:458]: Isn't proxy.prefix a byte array? Why do you need the `& 0xFF`?

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (fortuna) [inline cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:181]: Fail is the character is > 0xFF

File: cordova-plugin-outline/apple/src/OutlineTunnel.swift
Comment (fortuna) [inline cordova-plugin-outline/apple/src/OutlineTunnel.swift:26]: Should this be a byte array instead?
It's usually better to parse things as soon as possible.

File: cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m
Comment (fortuna) [inline cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m:281]: We should fail is character > 0xFF

File: src/electron/go_vpn_tunnel.ts
Comment (fortuna) [inline src/electron/go_vpn_tunnel.ts:244]: I don't think you need to encode anything.
I think we are spawning the process with a list of arguments, not with a command-line.
Let's remove the encoding if it works.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java:132]: Added a TODO

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:267]: Added a TODO on `connectTunnel`

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java:466]: Otherwise, the Java cast will interpret 0x80 as -128 and sign-extend it into 0xFF80.  I added a comment.

File: src/electron/go_vpn_tunnel.ts
Comment (bemasc) [reply src/electron/go_vpn_tunnel.ts]: OK, I removed this encoding.  I confirmed that Node uses UTF-8 to encode `child_process` args, so this means that the Go CLI utility has to un-UTF-8 those arguments.  I updated https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/98 with that logic.

File: cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m
Comment (bemasc) [reply cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m]: I moved this into Swift as suggested.

File: cordova-plugin-outline/apple/src/OutlineTunnel.swift
Comment (bemasc) [reply cordova-plugin-outline/apple/src/OutlineTunnel.swift]: OK, done.  I was not able to see a way for this parser to fail, so the parser tolerates arbitrary codepoints and silently ignores the high byte.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java]: This try-catch provides that check, as with `name` above.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java]: OK, done.

File: src/electron/go_vpn_tunnel.ts
Comment (fortuna) [reply src/electron/go_vpn_tunnel.ts]: I don't think we need to care about encodings.
We have two formats for the prefix: string or byte[].
If it's a string, we iterate over the (unicode) characters, using the unicode codepoint as the byte value.
At that point there's no encoding, the String implementation is returning numbers (unicode characters) to us.

**Review by fortuna — APPROVED** [review]
I think this is in good enough shape that we can submit and release for broader testing.

File: src/electron/go_vpn_tunnel.ts
Comment (bemasc) [reply src/electron/go_vpn_tunnel.ts]: Please look [here](https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/98/files#diff-af670972d38354924ac59bace6854c98e8379fb42753d4f6be70922fb9b70ae2R108-R124) to see the conversion that is required.

Strings in Javascript, Java, and Swift are (more or less) Unicode natively.  However, the nodejs `child_process.spawn()` functions pass arguments to the child process in UTF-8.  Go strings are byte sequences, not unicode-aware, so we start with a Go string containing UTF-8 bytes.  Then we need to convert from UTF-8 to Unicode codepoints (which Go calls "runes") and extract their numeric values.

I think this is all fine.

**Comment by bemasc** [comment]
> I think this is in good enough shape that we can submit and release for broader testing.

Thanks.  FYI, this is blocked on https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/98.

**Comment by fortuna** [comment]
Approved the other PR as well!

**Comment by bemasc** [comment]
Note: This PR still has the Android APK in it.  That needs to be deleted before squash-and-merge to avoid polluting `master` with wasteful binaries.

**Review by fortuna — APPROVED** [review]
Let's remove the APK before submitting. Squashing sounds good.

**Pull request merged**
